src/System.Tuples/src/System/ValueTuple/ValueTuple.cs

```diff
@@ -0,0 +1,1119 @@
+            new ValueTuple<T1, T2, T3, T4>(item1, item2, item3, item4);
+
+        public static ValueTuple<T1, T2, T3, T4, T5> Create<T1, T2, T3, T4, T5>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5) =>
+            new ValueTuple<T1, T2, T3, T4, T5>(item1, item2, item3, item4, item5);
+
+        public static ValueTuple<T1, T2, T3, T4, T5, T6> Create<T1, T2, T3, T4, T5, T6>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6) =>
+            new ValueTuple<T1, T2, T3, T4, T5, T6>(item1, item2, item3, item4, item5, item6);
+
+        public static ValueTuple<T1, T2, T3, T4, T5, T6, T7> Create<T1, T2, T3, T4, T5, T6, T7>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6, T7 item7) =>
+            new ValueTuple<T1, T2, T3, T4, T5, T6, T7>(item1, item2, item3, item4, item5, item6, item7);
+
+        public static ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest> Create<T1, T2, T3, T4, T5, T6, T7, TRest>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6, T7 item7, TRest rest) =>
+            new ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>(item1, item2, item3, item4, item5, item6, item7, rest);
+
+        // From System.Web.Util.HashCodeCombiner
+        internal static int CombineHashCodes(int h1, int h2)
+        {
+            return (((h1 << 5) + h1) ^ h2);
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2), h3);
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3, int h4)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2), CombineHashCodes(h3, h4));
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3, h4), h5);
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5, int h6)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3, h4), CombineHashCodes(h5, h6));
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5, int h6, int h7)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3, h4), CombineHashCodes(h5, h6, h7));
+        }
+
+        internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5, int h6, int h7, int h8)
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3, h4), CombineHashCodes(h5, h6, h7, h8));
+        }
+    }
+
+    public struct ValueTuple<T1> : IEquatable<ValueTuple<T1>>, IStructuralEquatable, IStructuralComparable, IComparable, ITupleInternal
+    {
+        public T1 Item1;
+
+        private static readonly EqualityComparer<T1> s_equality1 = EqualityComparer<T1>.Default;
+        private static readonly Comparer<T1> s_comparer1 = Comparer<T1>.Default;
```

Comment:
@jkotas, thanks.  The original implementation here had a field on ```ValueTuple<T1>``` that was then used by each of the Equals methods on the various types, e.g.
```C#
return 
    ValueTuple<T1>.s_comparer.Equals(...) &&
    ValueTuple<T2>.s_comparer.Equals(...) &&
    ...
```
Is that significantly different from a JIT perspective than using EqualityComparer<T1>.Default instead?

I tried the following microbenchmark (which I of course could have messed up, so please let me know if I did).  It has three implementations: a ```ValueTuple_SharedField<T1, T2>``` that uses a single cached ```ValueTuple<T>.s_comparer``` field, a ```ValueTuple_LocalField<T1, T2>``` that has its own set of cached fields for each of T1 and T2, and a ```ValueTuple_Default<T1, T2>``` that uses ```EqualityComparer<T>.Default```.
```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        var sw = new Stopwatch();
        var vtf1 = new ValueTuple_SharedField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtf2 = new ValueTuple_SharedField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtl1 = new ValueTuple_LocalField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtl2 = new ValueTuple_LocalField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtd1 = new ValueTuple_Default<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtd2 = new ValueTuple_Default<int, double> { Item1 = 42, Item2 = 42.0 };

        const int Iters = 100000000;
        bool result = true;
        while (true)
        {
            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtf1.Equals(vtf2);
            sw.Stop();
            Console.WriteLine($"Shared:  {sw.Elapsed.TotalSeconds}");

            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtl1.Equals(vtl2);
            sw.Stop();
            Console.WriteLine($"Local:   {sw.Elapsed.TotalSeconds}");

            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtd1.Equals(vtd1);
            sw.Stop();
            Console.WriteLine($"Default: {sw.Elapsed.TotalSeconds}");

            Console.WriteLine();
        }
    }
}

struct ValueTuple<T1> { internal static EqualityComparer<T1> s_comparer = EqualityComparer<T1>.Default; }

struct ValueTuple_SharedField<T1, T2> : IEquatable<ValueTuple_SharedField<T1, T2>>
{
    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_SharedField<T1, T2> other) =>
        ValueTuple<T1>.s_comparer.Equals(Item1, other.Item1) &&
        ValueTuple<T2>.s_comparer.Equals(Item2, other.Item2);
}

struct ValueTuple_LocalField<T1, T2> : IEquatable<ValueTuple_LocalField<T1, T2>>
{
    internal static readonly EqualityComparer<T1> s_comparer1 = EqualityComparer<T1>.Default;
    internal static readonly EqualityComparer<T2> s_comparer2 = EqualityComparer<T2>.Default;

    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_LocalField<T1, T2> other) =>
        s_comparer1.Equals(Item1, other.Item1) &&
        s_comparer2.Equals(Item2, other.Item2);
}

struct ValueTuple_Default<T1, T2> : IEquatable<ValueTuple_Default<T1, T2>>
{
    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_Default<T1, T2> other) =>
        EqualityComparer<T1>.Default.Equals(Item1, other.Item1) &&
        EqualityComparer<T2>.Default.Equals(Item2, other.Item2);
}
```
With the current mscorlib bits in coreclr, I get repeatable results like this, with each line varying only a little bit:
```
Shared:  0.9470538
Local:   0.6117238
Default: 2.1711197
```
So, the Default case is about 2x the Shared, and 3-4x the Local. However, if I change the mscorlib implementaton of Default to just pre-cache the comparer, rather than using a volatile field and lazy initialization, I get results like this:
```
Shared:  0.9644405
Local:   0.5083141
Default: 0.9415915
```
which suggests that the main difference between using a shared field and ```EqualityComparer<T>.Default``` is the implementation of Default, and in order to get the gains mentioned regarding caching a local comparer (which, per Jan's comments, avoids the generic type hop), we'd actually need to have a comparer field for each T1, T2, etc. on each ValueTuple type, which has the potential to result in a huge number of fields, depending on all the different combinations of types as generic type parameters to the various arities of ```ValueTuple<...>```.

As such, if that's all close to correct, seems to me that if this is a difference we care about, the answer is to change the implementation of ```EqualityComparer<T>.Default```, e.g. https://github.com/dotnet/coreclr/issues/1157, rather than adding such caching here.